tiledb/sm/query/readers/attribute_order_validator.h

```diff
@@ -0,0 +1,581 @@
+              rt->tile_idx() + frag_first_array_tile_idx[rt->frag_idx()],
+              dim_dom[0],
+              tile_extent) -
+          1;
+      AttributeType value_previous =
+          rt->attribute_value<AttributeType>(attribute_name_, cell_idx);
+
+      // Validate the order.
+      if (increasing_data) {
+        if (value_previous >= value) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      } else {
+        if (value_previous <= value) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      }
+    }
+
+    if (!max_validated_[f]) {
+      // Get the min of the current fragment.
+      auto max_tile_idx = fragment_metadata[f]->tile_num() - 1;
+      auto value = fragment_metadata[f]->get_tile_max_as<AttributeType>(
+          attribute_name_, max_tile_idx);
+
+      // Get the previous value from the loaded tile.
+      auto rt = max_tile_to_compare_against(f);
+      const auto cell_idx =
+          non_empty_domain[1] -
+          index_dim->tile_coord_low(
+              rt->tile_idx() + frag_first_array_tile_idx[rt->frag_idx()],
+              dim_dom[0],
+              tile_extent) +
+          1;
+      AttributeType value_next =
+          rt->attribute_value<AttributeType>(attribute_name_, cell_idx);
+
+      // Validate the order.
+      if (increasing_data) {
+        if (value >= value_next) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      } else {
+        if (value <= value_next) {
+          throw AttributeOrderValidatorStatusException(
+              "Attribute out of order");
+        }
+      }
+    }
+  }
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** Name of the attribute to validate. */
+  std::string attribute_name_;
+
+  /** Map of result tiles to load, per fragments. */
+  std::vector<std::unordered_map<uint64_t, ResultTile>> result_tiles_to_load_;
+
+  /** Vector to store, per fragment, which min has been validated already. */
+  std::vector<uint8_t> min_validated_;
+
+  /** Vector to store, per fragment, which max has been validated already. */
+  std::vector<uint8_t> max_validated_;
+
+  /**
+   * Fragments to compare against, per fragments. This will contain a value for
+   * both the min and the max bound of a fragment. If it was possible to
+   * validate the value without looking at another fragment, this will be
+   * nullopt.
+   */
+  std::vector<std::pair<optional<uint64_t>, optional<uint64_t>>>
+      fragment_to_compare_against_;
+
+  /**
+   * Vector of pairs to store, per fragment, which tile we should compare data
+   * against for the min/max. The value is an index into
+   * `result_tiles_to_load_`.
+   */
+  std::vector<std::pair<uint64_t, uint64_t>> tile_to_compare_against_;
```

Comment:
Should we have a struct that stores all the data needed for a single fragment rather than 4 separate vectors? 

I don't think this would be worth the development effort, but I think it is at least worth considering.